### ISSUE TITLE
Not compatible with FF 49

### ISSUE THREAD

**Issue by tobiasboyd** [jfitzell/mozilla-keychain#83]
See https://addons.mozilla.org/en-US/firefox/addon/keychain-services-integration/reviews/801966/

I don't have any experience with FF extensions, or I'd try to come up with a fix.

**Comment by jfitzell** [comment]
Thanks for raising. Anyone looking at fixing, see also Issue #36

**Comment by sierkb** [comment]
This functionality provided through his add-on is essentially for me using Firefox on Mac.
I have to use another browser which has got this functionality, if this doesn't function.
Please fix! As soon as possible.

**Comment by jfitzell** [comment]
I'm swamped at the moment I'm afraid and don't know when I'll have time to look at it. Obviously I'd like it fixed as much as anybody...

If any users of the extension have the skills to jump in and help, that would be terrific.

**Comment by Antwan86** [comment]
@jfitzell Can you give hints on where to implement the method (links/mozdev/whatever) ? It's pretty much a blackbox :/

**Comment by jfitzell** [comment]
Sure. The interface Mozilla cares about is all implemented in components/MacOSKeychainStorage.js

The API docs are here: https://developer.mozilla.org/en-US/docs/XPCOM_Interface_Reference/nsILoginManagerStorage

I think the main issue is that searchLogins() needs to be implemented. It used to get called and if it failed then Firefox would fall back on findLogins(), which was the original interface. Now searchLogins() seems to have become mandatory.

It might be possible to implement searchLogins() as a call to findLogins(), which would probably be the simplest thing to do if it works. If not, the nsiLoginManagerStorage implementation is a pretty thin layer in front of the Keychain-specific code, which is all in the modules/ directory. The main file in there is modules/KeychainServices.jsm, which wraps everything up into an interface that can be used by the storage component. You can either call one or more of the find*() methods directly or (if necessary) extend the functionality, but I don't think that should be necessary.

**Comment by bkimmett** [comment]
I just made a fix, and it is a call to findLogins(). Pull request is up.

**Comment by jfitzell** [comment]
@bkimmett Thanks very much for the fix... I have a new job, a new house, and a new baby so time is pretty tight at the moment. I merged and spent some time this morning testing. Looking up and adding passwords seem to work great but I was running into problems with updating passwords for existing logins - do the following work for you?

1. Signing in to a form with an existing account but new password: doorhanger asks if I want to update but incorrectly shows the old password value; if I retype the new value though, pressing `Update` correctly updates the password
2. Signing in to basic auth with an existing account but new password: doorhanger asks if I want to update and shows the new password value but incorrectly shows `Remember` as the button label; clicking the button fails with `Error: can't add a login without a httpRealm or formSubmitURL`. The logs show a call to `searchLogins(nsIPropertyBag{httpRealm: null, hostname: 'http://localhost', formSubmitURL: null, schemeUpgrades: true})` but that makes little sense since, as the error says, those values can't both be `null`.

Not sure whether those are bugs in the new FF49+ implementation of the password manager or if they're somehow caused by the implementation of `searchLogins()`, but would be good to be sure before doing a release.

Re: your comment about wildcard behaviour, I have the same interpretation as you: properties that are not provided must be wildcards. I suspect, though, that an empty string should match only an empty string, rather than being a wildcard as it is when passed to `findLogin()`. I see an empty string being passed in for `formSubmitURL` in some of my test cases, so I wonder if that's related somehow?

I added some more detailed debugging of the `nsIPropertyBags` that are passed into `searchLogins()` and `modifyLogin()` but wasn't able to spot the problem immediately. I'd obviously appreciate your help if you have time.

Would also be good to log any unsupported properties passed into `searchLogins()` (similar to what's done in `modifyLogin()` as it makes it easier to see what's going on when people report bugs and provide the logs.

**Comment by jfitzell** [comment]
I just tested quickly with the extension disabled and the two test cases I mentioned seem to work fine, so it does seem to be caused by the extension. Not sure whether it's `searchLogins()` causing the problem or some other change in the FF code though...

**Comment by bkimmett** [comment]
For test case 1, I can’t actually tell. My doorhanger doesn’t show passwords. If you know how to get it to show them, instructions would be much appreciated.

For test case 2, it sounds like the browser is misapprehending basic auth as being part of a local page (I’m assuming you tested it on basic auth that’s on the web somewhere).
The thing is, I was never entirely sure about which bit of code in Firefox calls searchLogins (I’m about 50% sure); If I had to guess, it’s a bug upstream with basic auth (I don’t have a basic auth page to test with, though).

What pages are you using for testing?

Ben K

**Comment by jfitzell** [comment]
Are you using the latest FF Nightly? Pretty sure it has a "show password" checkbox in the doorhanger, but I was just observing that the password length was wrong and that pressing "Update" calls `modifyLogin()` but doesn't pass in a new password to the method.

I use [these pages](https://github.com/jfitzell/mozilla-keychain/files/428797/keychain-test.zip) for quick and dirty testing - they need to be served over an HTTP server (i.e. `file://` URLs won't work). You need to configure Apache to `AllowOverrides` for the folder with those files in it, so that the .htaccess file can apply the basic auth restrictions. 

In terms of senders, you can search in DXR:
- The storage module's `searchLogins()` is called by the `LoginManager`'s `searchLogins()` here: https://dxr.mozilla.org/mozilla-central/source/toolkit/components/passwordmgr/nsLoginManager.js
- That's called by `searchLoginsWithObject()` in `LoginHelper`: https://dxr.mozilla.org/mozilla-central/source/toolkit/components/passwordmgr/LoginHelper.jsm
- that has a bunch of callers: https://dxr.mozilla.org/mozilla-central/search?q=%2Bref%3A%22LoginHelper%23searchLoginsWithObject%22

**Comment by bkimmett** [comment]
Oh, I see. I primarily use Firefox Aurora (but not Nightly). 
And thanks for the tip about the DXR. Someone directed me there at the start of my implementation period, but I was never quite sure if I was on the right searchLogins() call.

I’ll poke around for a day or two and see what might be the cause.

Ben K

**Comment by bkimmett** [comment]
Some quality time with the Browser Toolbox (you can debug Firefox from Firefox - how cool is that?!) is showing a hint of a problem source: when findLogins (in the extension) pulls test logins with no httpRealm set, after convertKeychainItemsToLoginInfos() is done, the returned ‘password’ value of the resulting login… is empty. I am still investigating and will try to find the exact circumstances that trigger an empty password return.

BK

**Comment by bkimmett** [comment]
…OK, so I missed how the whole business with the passwordHandler works. But I can’t see the passwordHandler proxy in the returned data from findLogins. Investigation continues.

**Comment by bkimmett** [comment]
OK, this is _really_ weird. MacOSKeychain.convertKeychainItemToLoginInfo() returns, for an individual login, a Proxy with a custom 'get' handler that pulls out the password on request. 
Then, on return from MacOSKeychain.convertKeychainItemsToLoginInfos() (into findLogins), this Proxy somehow turns into a pumpkin: that is, into a XPCWrappedNative_NoHelper with no apparent trace of the attached handler. However, the handler is still there, as when I try to do "logins[0].password", I get the correct password.

Could it be that the resulting code (after searchLogins returns) isn't using the getter returned with the object? This would point to a change in the FF49+ implementation of the password manager.

**Comment by jfitzell** [comment]
Some of that stuff is implemented by Firefox in native code so it's all a
bit magic how it works. I wouldn't be at all surprised if the debugger
doesn't quite show reality. So I *doubt* that's the problem, particularly
since fetching the passwords is working fine, but obviously I can't say for
sure.

**Comment by bkimmett** [comment]
I found the problem. 

Basically, the doorhanger generation function, _showChangeLoginNotification() in nsLoginManagerPrompter.js:1178, has this code in it:

`if (aNotifyObj == this._getPopupNote()) {
      aOldLogin.hostname = aNewLogin.hostname;
      aOldLogin.formSubmitURL = aNewLogin.formSubmitURL;
      aOldLogin.password = aNewLogin.password;
      aOldLogin.username = aNewLogin.username;
      this._showLoginCaptureDoorhanger(aOldLogin, "password-change");
    }`

Anyway, you set up a custom getter for your returned keychain login object, but I don't think there's a custom setter!! So the custom getter overrides it and aOldLogin's original password is pulled.

I think this is what is happening, anyway.

The correct solution is to check if it's safe to junk those first four lines of code inside the if block and just call:

`this._showLoginCaptureDoorhanger(aNewLogin, "password-change");`

...and then just submit a patch (for Firefox) to Mozilla if this is the case.
If we're going this way, it's worth noting that the doorhanger doesn't take the username and password values passed into it for storage— it takes the values from the displayed fields after the user's edited those fields (if they do), as is both observed and expected behavior.

The hacky solution is to modify your returned login object so it doesn't use a Proxy for its password and instead just stores it the normal way. I, uh, don't know how secure this is, though.

**Comment by jfitzell** [comment]
It's not just a security concern. The main reason for the proxy is to (try
to) avoid the user being prompted for access to every matching password as
soon as a page loads. We only want the prompt once they've decided they
want to login and selected which account they want to login with.

I say try to because when there are multiple password returned, Mozilla
checks the length of all the passwords against the length restrictions in
the passwords field, which forces all the passwords to be pulled in. I
asked them to remove that check but had no luck.

Your diagnosis sounds plausible, though I don't have the code to hand at
the moment to check what the getter/proxy look like. If they've started
modifying a returned LoginInfo and passing it back to us, and if there's no
setter, that does sound problematic. What if you add one?

**Comment by artmcgee** [comment]
"Signing in to a form with an existing account but new password: doorhanger asks if I want to update but incorrectly shows the old password value; if I retype the new value though, pressing Update correctly updates the password"

That's NOT a new problem. It has existed forever and I've almost ALWAYS had to manually paste the new password to make sure it updates. I clear all cookies each time my browser closes, so I depend on KSI more than the average person, which is why I may be aware of that while others haven't noticed it.

**Comment by oneofthedamons** [comment]
With Firefox 49 now released, I downgraded to Firefox 45 via the ESR channel today. Good to see this extension still works. Looks like I have until Q2 2017 before crunch time comes: https://www.mozilla.org/en-US/firefox/organizations/faq/

**Comment by tonglil** [comment]
I'm not sure it does work. I was trying to sign on to AWS today and it no longer autofills the credentials.

I'll dig into this when I get a chance.

**Comment by bkimmett** [comment]
Would it be possible to get a signed version of the extension with the searchLogins() patch out to Mozilla, and focus on resolving the second doorhanger bug afterwards? As some others have noted above, FF 49 is out to the general public, and this extension needs an update lest it be branded as 'really not working'.

**Comment by jfitzell** [comment]
Fair point... Makes sense

**Comment by tonglil** [comment]
That would be greatly appreciated!

**Comment by skyfrog** [comment]
please release new version with this fix for ff49.

**Comment by foocp** [comment]
completely understand time issues especially with new baby around  – still would love to have this working again!

**Comment by tsh69** [comment]
Is there a way to donate money for development?

**Comment by rolfmeijer** [comment]
> Is there a way to donate money for development?

Great idea, I would love to donate, as this is a key add-on (pun intended).

Groet’n
Rolf

**Comment by jirutka** [comment]
Me too, this add-on is essential.

**Comment by jfitzell** [comment]
Sorry for the delay. I've submitted 1.1.10 to Mozilla so it will be on the Addons site as soon as they approve it.

Have opened Issue 86 for the updating password problem.

**Comment by posita** [comment]
🎉 Thanks @jfitzell! Please let me (us) know where to send donations to the project!

**Comment by jfitzell** [comment]
@posita Thank you for the offer to donate. I've occasionally toyed with accepting donations, but always end up deciding against it. I created the extension out of a combination of frustration and curiosity, which makes me happy to give it away for free.

My challenge with updating it at the moment is time, which means money wouldn't be a motivation and it would just increase the guilt when I'm not able to fix things quickly enough. The most valuable thing would be a few other people helping to keep it updated — barring that, I'll just do what I can.

But again, thank you: it's nice to know that people find it useful!

**Comment by jfitzell** [comment]
v1.1.10 approved on AMO! https://addons.mozilla.org/en-US/firefox/addon/keychain-services-integration/

**Comment by skyfrog** [comment]
thank you!

**Comment by posita** [comment]
> My challenge with updating it at the moment is time, which means money wouldn't be a motivation and it would just increase the guilt when I'm not able to fix things quickly enough.

@jfitzell, my apologies. What I _really_ meant by my [original comment](https://github.com/jfitzell/mozilla-keychain/issues/83#issuecomment-252455854) was: please let us know how we can best express our gratitude. I didn't mean to try to obligate you. Let me try again.

I have a history of dissatisfaction with web browsers. I have been Firefox loyal for awhile, in no small part because I'm creeped out by the spyware in Chrome and annoyed by the quirkiness of Safari. I refuse to even engage the (often purposefully) broken IE (it feels like 20-50% of a UI developer's work in supporting multiple browsers goes to dealing with the absolutely :poop:-show that is IE). But one very, very useful feature of both Safari and Chrome is their respective abilities to integrate with OS X's keychain. Your work brings that to Firefox in such a seamless way, that until it stopped working, I forgot it was not a native feature. It made me realize how much time I've saved, and joy I experienced through your work. When it threatened to go away, the anxiety I experienced was not trivial. I am very grateful to you for taking the time to help make it work again. Thank you.

> The most valuable thing would be a few other people helping to keep it updated — barring that, I'll just do what I can.

Let me second this. I know hardly anything about Mozilla add-on development, but perhaps issues like these are good excuses to learn (for me, and software developers like me who happen to benefit from your work).

**Comment by oneofthedamons** [comment]
👍

**Comment by mysheepb** [comment]
just want to thank @bkimmett and @jfitzell for fixing this absolutely essential add-on!

**Comment by rolfmeijer** [comment]
Thank you for updating, it’s working like a charm again and the previous week reminded me what a lovely add-on this is.

> My challenge with updating it at the moment is time, which means money wouldn't be a motivation and it would just increase the guilt when I'm not able to fix things quickly enough. The most valuable thing would be a few other people helping to keep it updated — barring that, I'll just do what I can.

Donation would be an expression of gratitude for proven services, not an expectation for coming updates. But I see your point. :-)

**Comment by foocp** [comment]
Thank you! 👍
